tests/taskgraphs/test_client_executor.py

```diff
@@ -86,6 +88,62 @@ def test_failure(self):
         self.assertIs(executor.Status.FAILED, exec.status)
 
 
+class ClientExecutorTestLocal(unittest.TestCase):
+    def test_basic(self):
+        grf = builder.TaskGraphBuilder(name="test_local")
+        double = grf.udf(lambda x: 2 * x, types.args(2), local=True)
+
+        exec = client_executor.LocalExecutor(grf, name="test_local exec")
+        exec.execute()
+        double_exec = exec.node(double)
+        self.assertEqual(4, double_exec.result(30))
+        self.assertIsNone(double_exec.task_id())
+        exec.wait(10)
+
+    def test_concurrent(self):
+        import os
+        import os.path
+        import time
+
+        def touch_and_wait(dir, file):
+            full_path = os.path.join(dir, file)
+            open(full_path, "w").close()
+            while os.path.exists(full_path):
+                time.sleep(0.01)
+            return file
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            grf = builder.TaskGraphBuilder()
+            one = grf.udf(touch_and_wait, types.args(tmpdir, "one"), local=True)
+            two = grf.udf(touch_and_wait, types.args(tmpdir, "two"), local=True)
+            out = grf.udf(" ".join, types.args((one, two)), local=True)
+            exec = client_executor.LocalExecutor(grf)
+            exec.execute()
+            with self.assertRaises(futures.TimeoutError):
+                exec.node(one).result(1)
+            with self.assertRaises(futures.TimeoutError):
+                exec.node(two).result(1)
+
+            p_one = os.path.join(tmpdir, "one")
+            p_two = os.path.join(tmpdir, "two")
+
+            while not os.path.exists(p_one):
+                time.sleep(0.01)
+            while not os.path.exists(p_two):
+                time.sleep(0.01)
```

Comment:
Are these `sleep` calls necessary? IIRC the files must be created after line 125

Reply:
Not quite; as of lines 122–125 the `touch_and_wait` tasks are stuck on their own sleep-loop (lines 111–112) waiting for the files to be removed. I’ve added some more comments explaining the sequencing of the test.

Reply:
>  the touch_and_wait tasks are stuck on their own sleep-loop (lines 111–112) waiting for the files to be removed

Right, but before entering their sleep loop they create the files. This implies `os.path.exists(p_one) and os.path.exists(p_two)` after the timeout errors, in which case the `sleep()` at lines 131 and 133 are never called, no? (sorry if I'm still missing it).

Reply:
While unlikely, the tasks may take a little while to spin up, in which case even after 1 second the creation may not have yet occurred. Waiting here to ensure that the files *do* exist guarantees the correctness of the test.